src/EFCore.PG.NTS/Storage/Internal/NpgsqlNetTopologySuiteTypeMappingSourcePlugin.cs

```diff
@@ -14,23 +15,92 @@ public class NpgsqlNetTopologySuiteTypeMappingSourcePlugin : IRelationalTypeMapp
-                        storeTypeName.Equals("geography", StringComparison.OrdinalIgnoreCase))
+            if (clrType != null && !typeof(IGeometry).IsAssignableFrom(clrType))
+                return null;
+
+            if (storeTypeName != null)
+            {
+                if (!TryParseStoreTypeName(storeTypeName, out isGeography, out var parsedSubtype, out var _))
+                    return null;
+                if (clrType == null)
+                    clrType = parsedSubtype;
+            }
+
+            return clrType != null || storeTypeName != null
                 ? (RelationalTypeMapping)Activator.CreateInstance(
                     typeof(NpgsqlGeometryTypeMapping<>).MakeGenericType(clrType ?? typeof(IGeometry)),
-                    storeTypeName ?? (_options.IsGeographyDefault ? "geography" : "geometry"))
+                    storeTypeName ?? (isGeography ? "geography" : "geometry"),
+                    isGeography)
                 : null;
         }
+
+        /// <summary>
+        /// Given a PostGIS store type name (e.g. GEOMETRY, GEOGRAPHY(Point, 4326)), attempts to parse it and return its components.
+        /// </summary>
+        public static bool TryParseStoreTypeName(string storeTypeName, out bool isGeography, out Type clrType, out int srid)
+        {
+            isGeography = false;
+            clrType = null;
+            srid = -1;
+
+            var openParen = storeTypeName.IndexOf("(", StringComparison.Ordinal);
+
+            var baseType = (openParen > 0
+                ? storeTypeName.Substring(0, openParen)
+                : storeTypeName)
+                .Trim();
+
+            if (baseType.Equals("geometry", StringComparison.OrdinalIgnoreCase))
+                isGeography = false;
+            else if (baseType.Equals("geography", StringComparison.OrdinalIgnoreCase))
+                isGeography = true;
+            else
+                return false;
+
+            if (openParen == -1)
+                return true;
+
+            string subTypeString = null;
+            var closeParen = storeTypeName.IndexOf(")", openParen + 1, StringComparison.Ordinal);
+            if (closeParen > openParen)
+            {
+                var comma = storeTypeName.IndexOf(",", openParen + 1, StringComparison.Ordinal);
+                if (comma > openParen && comma < closeParen)
+                {
+                    subTypeString = storeTypeName.Substring(openParen + 1, comma - openParen - 1).Trim();
+
+                    if (!int.TryParse(storeTypeName.Substring(comma + 1, closeParen - comma - 1).Trim(), out srid))
+                        return false;
+                }
+                else
+                    subTypeString = storeTypeName.Substring(openParen + 1, closeParen - openParen - 1).Trim();
+            }
+
+            if (subTypeString != null && !SubTypeNameToClrType.TryGetValue(subTypeString, out clrType))
+                return false;
+
+            return true;
```

Comment:
Then this four lines could be replaced by a single method call on the dictionary.

Reply:
Done